commit 45a55ffb918bb57842ffa8972e931b991b524237
Author: Susan Franklin <sfranklin>
Date:   Wed Dec 10 00:03:15 2008 +0000

    Bug 258132 - [ui] I can install something while an installation is proceeding

3	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
1	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
7	2	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
7	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
1	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
1	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/IStatusCodes.java
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
index 52fbe36..9168680 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
@@ -46,7 +46,9 @@ public class PlanStatusHelper {
 			case IStatusCodes.PROFILE_CHANGE_ALTERED :
 				return new MultiStatus(ProvUIActivator.PLUGIN_ID, statusCode, ProvUIMessages.PlanStatusHelper_RequestAltered, null);
 			case IStatusCodes.NOTHING_TO_UPDATE :
-				return new Status(IStatus.INFO, ProvUIActivator.PLUGIN_ID, IStatusCodes.NOTHING_TO_UPDATE, ProvUIMessages.UpdateOperation_NothingToUpdate, null);
+				return new Status(IStatus.INFO, ProvUIActivator.PLUGIN_ID, statusCode, ProvUIMessages.UpdateOperation_NothingToUpdate, null);
+			case IStatusCodes.OPERATION_ALREADY_IN_PROGRESS :
+				return new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, statusCode, ProvUIMessages.PlanStatusHelper_AnotherOperationInProgress, null);
 			default :
 				return new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, statusCode, NLS.bind(ProvUIMessages.PlanStatusHelper_UnexpectedError, new Integer(statusCode), getIUString(affectedIU)), null);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
index 849dd84..b0208b5 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
@@ -176,6 +176,7 @@ public class ProvUIMessages extends NLS {
 	public static String PlanStatusHelper_Items;
 	public static String PlanStatusHelper_NothingToDo;
 	public static String PlanStatusHelper_AlreadyInstalled;
+	public static String PlanStatusHelper_AnotherOperationInProgress;
 	public static String PlanStatusHelper_Launch;
 	public static String PlanStatusHelper_RequestAltered;
 	public static String PlanStatusHelper_RequiresUpdateManager;
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
index ca9fb79..997e682 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
@@ -15,6 +15,7 @@ import java.util.HashSet;
 import org.eclipse.equinox.internal.p2.ui.model.ElementUtils;
 import org.eclipse.equinox.internal.p2.ui.model.IUElementListRoot;
 import org.eclipse.equinox.internal.provisional.p2.director.ProvisioningPlan;
+import org.eclipse.equinox.internal.provisional.p2.ui.ProvisioningOperationRunner;
 import org.eclipse.equinox.internal.provisional.p2.ui.policy.Policy;
 import org.eclipse.jface.wizard.IWizardPage;
 import org.eclipse.jface.wizard.Wizard;
@@ -99,7 +100,7 @@ public abstract class ProvisioningOperationWizard extends Wizard {
 	public IWizardPage getNextPage(IWizardPage page) {
 		if (page == mainPage) {
 			if (resolutionPage != null) {
-				if (mainPageSelectionsHaveChanged()) {
+				if (shouldRecomputePlan()) {
 					// any initial plan that was passed in is no longer valid, no need to hang on to it
 					plan = null;
 					planSelections = mainPage.getCheckedIUElements();
@@ -107,7 +108,7 @@ public abstract class ProvisioningOperationWizard extends Wizard {
 					planChanged();
 				}
 			} else {
-				if (plan != null && mainPageSelectionsHaveChanged())
+				if (plan != null && shouldRecomputePlan())
 					plan = null;
 				resolutionPage = createResolutionPage(makeResolutionElementRoot(mainPage.getCheckedIUElements()), plan);
 				planChanged();
@@ -118,6 +119,10 @@ public abstract class ProvisioningOperationWizard extends Wizard {
 		return null;
 	}
 
+	private boolean shouldRecomputePlan() {
+		return ProvisioningOperationRunner.hasScheduledOperationsFor(profileId) || mainPageSelectionsHaveChanged();
+	}
+
 	private boolean mainPageSelectionsHaveChanged() {
 		HashSet selectedIUs = new HashSet();
 		selectedIUs.addAll(Arrays.asList(ElementUtils.elementsToIUs(mainPage.getCheckedIUElements())));
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
index 0df8692..dc45966 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionWizardPage.java
@@ -267,6 +267,13 @@ public abstract class ResolutionWizardPage extends WizardPage {
 				// Log warnings for later support
 				ProvUI.reportStatus(currentStatus, StatusManager.LOG);
 			}
+		} else {
+			// Check to see if another operation is in progress
+			if (ProvisioningOperationRunner.hasScheduledOperationsFor(profileId)) {
+				messageType = IMessageProvider.ERROR;
+				currentStatus = PlanStatusHelper.getStatus(IStatusCodes.OPERATION_ALREADY_IN_PROGRESS, null);
+				pageComplete = false;
+			}
 		}
 		setPageComplete(pageComplete);
 		setMessage(getMessageText(), messageType);
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
index 2764059..dfba92d 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
@@ -145,6 +145,7 @@ PlanStatusHelper_ImpliedUpdate={0} is already installed, so an update will be pe
 PlanStatusHelper_Items=Items
 PlanStatusHelper_NothingToDo=Cannot complete the request.  See the details.
 PlanStatusHelper_AlreadyInstalled={0} will be ignored because it is already installed.
+PlanStatusHelper_AnotherOperationInProgress=Cannot continue the operation.  There is another install operation in progress.
 PlanStatusHelper_Launch=Launch...
 PlanStatusHelper_RequestAltered=Your original request has been modified.
 PlanStatusHelper_RequiresUpdateManager=Cannot continue the operation.  One or more items requires support from the old Update Manager.
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/IStatusCodes.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/IStatusCodes.java
index c2358e5..0a2582f 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/IStatusCodes.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/IStatusCodes.java
@@ -30,4 +30,5 @@ public interface IStatusCodes {
 	public static final int IGNORED_ALREADY_INSTALLED = 10004;
 	public static final int UNEXPECTED_NOTHING_TO_DO = 10005;
 	public static final int EXPECTED_NOTHING_TO_DO = 10006;
+	public static final int OPERATION_ALREADY_IN_PROGRESS = 10007;
 }
